File: Exercise_2.cpp
// Time Complexity                              :  O(N*M), N is the row size and M is column size of the given matrix
// Space Complexity                             :  O(1)
// Did this code successfully run on Leetcode   :  Yes
// Any problem you faced while coding this      :  No

#include <bits/stdc++.h> 
using namespace std;  


class Solution {
public:
    vector<int> findDiagonalOrder(vector<vector<int>>& mat) {
        int n = mat.size();
        int m = mat[0].size();
        
            int dir = 1;
            vector<int> ans(n*m);
            int index = 0;
            int r = 0, c = 0;
        
            while(index < m*n) {
                
                ans[index++] = mat[r][c];
                
                if(dir == 1) {            // moving in upward diagonal direction
                    if(c == n-1) {        // after hitting the last column, the row increases by 1 and direction reverses
                        r++;
                        dir = -1;
                    } else if (c == 0) {  // after hitting column 0, the column increases and direction reduces.
                        c++;
                        dir = -1;
                    } else {              // in normal case, we keep moving up and right
                        r--;
                        c++;
                    }
                } else {                    // moving in downward diagonal direction
                    if(r == m-1) {          // after hitting the last row, the column increases by 1 and direction reverses.
                        c++;
                        dir = 1;
                    } else if (r == 0) {   // after hitting the last row, row increases and direction reverses
                        r++;
                        dir = 1;
                    } else {               // in normal case, we keep moving down and left.
                        r++;
                        c--;
                    }
                }
           
            }
        
        return ans;
        
    }
};

File: Exercise_2.js
// Time Complexity : O(m * n) where m*n is matrix
// Space Complexity : O(m * n)
// Approach : transversing through matrix and handling border conditions with dir pointer which is true for upwards and false for downwards

/**
 * @param {number[][]} mat
 * @return {number[]}
 */
var findDiagonalOrder = function (mat) {
    // true upwards false downwards
    let dir = true
    let m = mat.length
    let n = mat[0].length
    let resArr = new Array(m * n)
    let i = 0
    let j = 0
    for (let idx = 0; idx < resArr.length; idx++) {
        resArr[idx] = mat[i][j]
        if (dir) {
            if (i == 0 && j != n - 1) {
                j++
                dir = false
            } else if (j == n - 1) {
                i++
                dir = false
            } else {
                i--
                j++
            }
        } else {
            if (j == 0 && i != m - 1) {
                i++
                dir = true
            } else if (i == m - 1) {
                j++
                dir = true
            } else {
                j--
                i++
            }
        }
    }
    return resArr
};